ipam/store.go

```diff
@@ -17,85 +17,130 @@ package ipam
 
 import (
 	"database/sql"
+	"errors"
+	"fmt"
 	"github.com/romana/core/common"
 	"log"
+	"strings"
 )
 
-type Vm struct {
-	//	Id        uint64 `json:"id"`
+type Endpoint struct {
 	Ip           string         `json:"ip"`
 	TenantId     string         `json:"tenant_id"`
 	SegmentId    string         `json:"segment_id"`
 	HostId       string         `json:"host_id"`
 	Name         string         `json:"instance"`
 	RequestToken sql.NullString `json:"request_token" sql:"unique"`
-	// Ordinal number of this VM in the host/tenant combination
-	Seq uint64 `json:"sequence"`
-	// Calculated effective sequence number of this VM --
+	// Ordinal number of this Endpoint in the host/tenant combination
+	Seq uint64 `json:"sequence",json:"-"`
+	// Calculated effective sequence number of this Endpoint --
 	// taking into account stride (endpoint space bits)
 	// and alignment thereof. This is used in IP calculation.
-	EffectiveSeq uint64 `json:"effective_sequence"`
+	EffectiveSeq uint64 `json:"effective_sequence",json:"-"`
+	// Whether it is in use (for purposes of reclaiming)
+	InUse bool   `json:"in_use",json:"-"`
+	Id    uint64 `sql:"AUTO_INCREMENT"`
 }
 
 type IpamHost struct {
-	Vms []IpamVm
-	Id  string `sql:"unique_index"`
+	Endpoints []Endpoint
+	Id        string `sql:"unique_index"`
 }
 
 type IpamSegment struct {
-	Vms []IpamVm
-	Id  string `sql:"unique_index"`
-}
-
-// TODO change this to Endpoint
-type IpamVm struct {
-	Vm
-
-	Id uint64 `sql:"AUTO_INCREMENT"`
-	//	IpamHostId sql.NullString
+	Endpoints []Endpoint
+	Id        string `sql:"unique_index"`
 }
 
 type ipamStore struct {
 	common.DbStore
 }
 
-func (ipamStore *ipamStore) addVm(stride uint, vm *Vm) error {
+func (ipamStore *ipamStore) deleteEndpoint(ip string) (Endpoint, error) {
 	tx := ipamStore.DbStore.Db.Begin()
+	results := make([]Endpoint, 0)
+	tx.Where(&Endpoint{Ip: ip}).Find(&results)
+	if len(results) == 0 {
+		tx.Rollback()
+		return Endpoint{}, common.NewError404("Endpoint", ip)
+	}
+	if len(results) > 1 {
+		// This cannot happen by constraints...
+		tx.Rollback()
+		errMsg := fmt.Sprintf("Expected one result for ip %s, got %v", ip, results)
+		log.Printf(errMsg)
+		return Endpoint{}, common.NewError500(errors.New(errMsg))
+	}
 
-	row := tx.Model(IpamVm{}).Where("host_id = ? AND segment_id = ?", vm.HostId, vm.SegmentId).Select("IFNULL(MAX(seq),-1)+1").Row()
-	row.Scan(&vm.Seq)
-	log.Printf("New sequence is %d\n", vm.Seq)
-
-	// vmSeq is the sequence number of VM in a given host
-	effectiveVmSeq := getEffectiveSeq(vm.Seq, stride)
-	log.Printf("Effective sequence for seq %d (stride %d): %d\n", vm.Seq, stride, effectiveVmSeq)
-	vm.EffectiveSeq = effectiveVmSeq
-	ipamVm := IpamVm{Vm: *vm}
-	tx.NewRecord(ipamVm)
-	tx = tx.Create(&ipamVm)
+	tx = tx.Model(Endpoint{}).Where("ip = ?", ip).Update("in_use", false)
```

Comment:
So, this function doesn't actually delete anything, it just marks things as not in use anymore? Might be worth a comment.

Are those things ever actually deleted?